sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -7857,4 +7859,156 @@ pi_result piTearDown(void *PluginParameter) {
   return PI_SUCCESS;
 }
 
+pi_result _pi_buffer::getZeHandlePtr(char **&ZeHandlePtr,
+                                     access_mode_t AccessMode,
+                                     pi_device Device) {
+  char *ZeHandle;
+  PI_CALL(getZeHandle(ZeHandle, AccessMode, Device));
+  ZeHandlePtr = &Allocations[Device].ZeHandle;
+  return PI_SUCCESS;
+}
+
+pi_result _pi_buffer::getZeHandle(char *&ZeHandle, access_mode_t AccessMode,
+                                  pi_device Device) {
+
+  if (!Device)
+    Device = LastDeviceWithValidAllocation;
+  PI_ASSERT(Device, PI_INVALID_DEVICE);
+
+  auto &Allocation = Allocations[Device];
+
+  // Sub-buffers don't maintain own allocations but rely on parent buffer.
+  if (isSubBuffer()) {
+    PI_CALL(SubBuffer.Parent->getZeHandle(ZeHandle, AccessMode, Device));
+    ZeHandle += SubBuffer.Origin;
+    // Still store the allocation info in the PI sub-buffer for
+    // getZeHandlePtr to work. At least zeKernelSetArgumentValue needs to
+    // be given a pointer to the allocation handle rather than its value.
+    //
+    Allocation.ZeHandle = ZeHandle;
+    LastDeviceWithValidAllocation = Device;
+    return PI_SUCCESS;
+  }
+
+  // Peform actual allocation as needed.
+  if (!Allocation.ZeHandle) {
+    // Choose an alignment that is at most 64 and is the next power of 2
+    // for sizes less than 64.
+    auto Alignment = Size;
+    if (Alignment > 32UL)
+      Alignment = 64UL;
+    else if (Alignment > 16UL)
+      Alignment = 32UL;
+    else if (Alignment > 8UL)
+      Alignment = 16UL;
+    else if (Alignment > 4UL)
+      Alignment = 8UL;
+    else if (Alignment > 2UL)
+      Alignment = 4UL;
+    else if (Alignment > 1UL)
+      Alignment = 2UL;
+    else
+      Alignment = 1UL;
+
+    if (OnHost) {
+      if (HostPtrImported) {
+        // When HostPtr is imported we use it for the buffer.
+        ZeHandle = MapHostPtr;
+      } else {
+        PI_CALL(piextUSMHostAlloc(pi_cast<void **>(&ZeHandle), Context, nullptr,
+                                  Size, Alignment));
+      }
+    } else {
+      if (Context->SingleRootDevice && Context->SingleRootDevice != Device) {
+        // If all devices in the context are sub-devices of the same device
+        // then we reuse root-device allocation by all sub-devices in the
+        // context.
+        // TODO: we can probably generalize this and share root-device
+        //       allocations by its own sub-devices even if not all other
+        //       devices in the context have the same root.
+        PI_CALL(getZeHandle(ZeHandle, AccessMode, Context->SingleRootDevice));
+      } else { // Create device allocation
+        PI_CALL(piextUSMDeviceAlloc(pi_cast<void **>(&ZeHandle), Context,
+                                    Device, nullptr, Size, Alignment));
+      }
```

Comment:
Before this patch we had control enableBufferPooling() which allowed to switch between pooling and non-pooling modes.
You remove this control in this patch. 
It would be great if @rdeodhar can confim that it is ok.

@smaslov-intel, since you removed enableBufferPooling() usage from plugin then could you please remove the declaration and defintion of this function from these files as well:
plugins/level_zero/usm_allocator.cpp:bool enableBufferPooling() { return USMSettings.EnableBuffers; }
plugins/level_zero/usm_allocator.hpp:bool enableBufferPooling();

Reply:
> Before this patch we had control enableBufferPooling() which allowed to switch between pooling and non-pooling modes. You remove this control in this patch. It would be great if @rdeodhar can confim that it is ok.

I retained the enableBufferPooling in this patch because it is orthogonal to the problem being fixed.